hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/list/ITestS3AReadRestoredGlacierObjects.java

```diff
@@ -0,0 +1,198 @@
+    Configuration conf = this.createConfiguration();
+    conf.set(READ_RESTORED_GLACIER_OBJECTS, s3ObjectStorageClassFilter);
+    conf.set(STORAGE_CLASS, glacierClass); // Create Glacier objects:Storage Class:DEEP_ARCHIVE/GLACIER
+    S3AContract contract = (S3AContract) createContract(conf);
+    contract.init();
+
+    FileSystem fs = contract.getTestFileSystem();
+    Path dir = methodPath();
+    fs.mkdirs(dir);
+    Path path = new Path(dir, "file1");
+    ContractTestUtils.touch(fs, path);
+    return fs;
+  }
+
+  @Override
+  protected Configuration createConfiguration() {
+    Configuration newConf = super.createConfiguration();
+    skipIfStorageClassTestsDisabled(newConf);
+    disableFilesystemCaching(newConf);
+    removeBaseAndBucketOverrides(newConf, STORAGE_CLASS);
+    newConf.set(REJECT_OUT_OF_SPAN_OPERATIONS, "false");
+    return newConf;
+  }
+
+  @Test
+  public void testIgnoreGlacierObject() throws Throwable {
+    Assume.assumeTrue(type == Type.GLACIER_AND_DEEP_ARCHIVE);
+    try (FileSystem fs = createFiles(S3ObjectStorageClassFilter.SKIP_ALL_GLACIER.name())) {
+      Assertions.assertThat(
+          fs.listStatus(methodPath()))
+        .describedAs("FileStatus List of %s", methodPath()).isEmpty();
+    }
+  }
+
+  @Test
+  public void testIgnoreRestoringGlacierObject() throws Throwable {
+    Assume.assumeTrue(type == Type.GLACIER_AND_DEEP_ARCHIVE);
+    try (FileSystem fs = createFiles(S3ObjectStorageClassFilter.READ_RESTORED_GLACIER_OBJECTS.name())) {
+      Assertions.assertThat(
+              fs.listStatus(
+                  methodPath()))
+          .describedAs("FileStatus List of %s", methodPath()).isEmpty();
+    }
+  }
+
+  @Test
+  public void testRestoredGlacierObject() throws Throwable {
+    Assume.assumeTrue(type == Type.GLACIER); // Skipping this test for Deep Archive as expedited retrieval is not supported
+    try (FileSystem fs = createFiles(S3ObjectStorageClassFilter.READ_RESTORED_GLACIER_OBJECTS.name())) {
+      restoreGlacierObject(methodPath().toUri().getHost(), getFilePrefixForListObjects() + "file1", 2);
+      Assertions.assertThat(
+              fs.listStatus(
+                  methodPath()))
+          .describedAs("FileStatus List of %s", methodPath()).isNotEmpty();
+    }
+  }
+
+  @Test
+  public void testDefault() throws Throwable {
+    Assume.assumeTrue(type == Type.GLACIER_AND_DEEP_ARCHIVE);
+    try (FileSystem fs = createFiles(S3ObjectStorageClassFilter.READ_ALL.name())) {
+      Assertions.assertThat(
+              fs.listStatus(methodPath()))
+          .describedAs("FileStatus List of %s", methodPath()).isNotEmpty();
+    }
+  }
+
+
+  private void restoreGlacierObject(String bucketName, String glacierObjectKey, int expirationDays) {
+
+    S3Client s3Client = getFileSystem().getS3AInternals().getAmazonS3Client("test");
+
+    // Create a restore object request
+    RestoreObjectRequest requestRestore = RestoreObjectRequest.builder()
```

Comment:
prefer this was in the RequestFactory interface and builder, as it'll let us do things like add audit context and anything else in future